File: netDxf/Objects/DxfObjectDatabase.Erase.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using netDxf.Blocks;
using netDxf.Entities;
using netDxf.Header;
using netDxf.IO;

namespace netDxf.Objects
{
    public sealed partial class DxfObjectDatabase
    {
        /// <summary>Permanently erases a registered object and every typed object owned by it.</summary>
        /// <param name="root">The registered root of the ownership subtree, including an object whose owning dictionary names were already removed.</param>
        /// <remarks>
        /// All owning aliases of the selected root and its reciprocal host extension attachment are removed.
        /// Other incoming references reject the operation before any mutation. Ownership follows common
        /// owners, not pointer strength; referenced entities, blocks and resources are never cascaded.
        /// Opaque objects inside the subtree reject. Surviving opaque exposed handle fields are checked,
        /// but dependencies hidden in private strings or binary data cannot be proven absent.
        /// Erasure is terminal: handles and payload remain available for inspection, without a resurrection API.
        /// </remarks>
        public void EraseOwnedTree(DxfDatabaseObject root)
        {
            if (root == null) throw new ArgumentNullException(nameof(root));
            if (root.IsErased) throw new InvalidOperationException("The object has already been erased.");
            if (root.Database != this) throw new ArgumentException("The object must belong to this database.", nameof(root));
            this.CheckRegistered(root);
            if (ReferenceEquals(root, this.Root)) throw new InvalidOperationException("The named object dictionary root cannot be erased.");

            // Snapshot all retained metadata carriers, including ATTRIBs that are not in AddedObjects.
            List<DxfObject> carriers = this.ErasureCarriers();
            var children = new Dictionary<DxfObject, List<DxfObject>>(ObjectIdentity);
            foreach (DxfObject item in carriers)
                if (item.Owner != null)
                {
                    if (!children.TryGetValue(item.Owner, out List<DxfObject> owned)) children.Add(item.Owner, owned = new List<DxfObject>());
                    owned.Add(item);
                }
            var deleted = new HashSet<DxfObject>(ObjectIdentity);
            var tree = new List<DxfDatabaseObject>();
            var pending = new Queue<DxfObject>(); pending.Enqueue(root);
            while (pending.Count != 0)
            {
                DxfObject item = pending.Dequeue();
                if (!deleted.Add(item)) throw new InvalidOperationException("The erased ownership subtree contains a cycle.");
                if (!(item is DxfDatabaseObject value) || ReferenceEquals(item, this.Root))
                    throw new NotSupportedException("An ownership subtree containing a managed legacy object cannot be erased.");
                if (value.IsErased || value.Database != this || !this.IsRegistered(value))
                    throw new InvalidOperationException("The erased ownership subtree has inconsistent registration.");
                if (value is DxfStoredDimAssoc) throw new NotSupportedException("Stored DIMASSOC erasure requires the complete dimension association lifecycle.");
                if (value is DxfOpaqueObject) throw new NotSupportedException("An opaque object requires its application schema before erasure: " + value.CodeName);
                tree.Add(value);
                if (children.TryGetValue(item, out List<DxfObject> next)) foreach (DxfObject child in next) pending.Enqueue(child);
            }
            if (root.Owner == null || !this.IsRegistered(root.Owner)) throw new InvalidOperationException("The erased root has no registered owner.");
            if (ReferenceEquals(root.Owner, this.Document.Layers)) throw new NotSupportedException("The LAYER table extension is managed by the layer-state collection.");
            DxfDictionary parent = root.Owner as DxfDictionary;
            var aliases = parent == null ? new List<DxfDictionaryEntry>() : parent.Entries.Where(e => ReferenceEquals(e.Target, root)).ToList();
            if (ReferenceEquals(parent, this.Root) && aliases.Any(e => IsReservedName(e.Name)))
                throw new NotSupportedException("Managed legacy dictionary entries cannot be erased through this API.");
            bool extension = ReferenceEquals(root.Owner.ExtensionDictionary, root);
            if (parent == null && !extension) throw new NotSupportedException("Erase an object through its owning dictionary or reciprocal extension attachment.");

            var handles = new HashSet<ulong>();
            foreach (DxfDatabaseObject item in tree)
            {
                if (!handles.Add(ErasureHandle(item.Handle))) throw new InvalidOperationException("The erased ownership subtree contains duplicate numeric handles.");
                // AddedObjects.Remove releases these counts and subscriptions. Verify its lookups now,
                // before the commit performs any collection mutation or event unsubscription.
                foreach (XData data in item.XData.Values)
                    if (!this.Document.ApplicationRegistries.References.ContainsKey(data.ApplicationRegistry.Name))
                        throw new InvalidOperationException("An erased object has inconsistent APPID reference bookkeeping: " + data.ApplicationRegistry.Name);
            }

            foreach (DxfObject item in carriers)
            {
                if (deleted.Contains(item)) continue;
                Action<DxfObject, string> reference = (target, field) =>
                {
                    if (target != null && deleted.Contains(target)) throw ErasureReference(item, field, target.Handle);
                };
                Action<string, string> handle = (target, field) =>
                {
                    if (handles.Contains(ErasureHandle(target))) throw ErasureReference(item, field, target);
                };
                reference(item.Owner, "owner");
                if (!(extension && ReferenceEquals(item, root.Owner) && ReferenceEquals(item.ExtensionDictionary, root)))
                    reference(item.ExtensionDictionary, "extension dictionary");
                foreach (DxfObject reactor in item.PersistentReactors) reference(reactor, "persistent reactor");
                if (item is EntityObject entity) foreach (DxfObject reactor in entity.Reactors) reference(reactor, "entity reactor");
                foreach (XData data in item.XData.Values)
                    foreach (XDataRecord tag in data.XDataRecord)
                        if (tag.Code == XDataCode.DatabaseHandle) handle((string)tag.Value, "XData 1005");
                if (item is DxfDatabaseObject databaseObject)
                    foreach (DxfObject target in databaseObject.DatabaseReferences) reference(target, "typed reference");
                if (item is DxfDictionary dictionary)
                    foreach (DxfDictionaryEntry entry in dictionary.Entries)
                        if (!(ReferenceEquals(dictionary, parent) && ReferenceEquals(entry.Target, root))) reference(entry.Target, "dictionary entry " + entry.Name);
                if (item is DxfDictionaryWithDefault fallback) reference(fallback.Default, "dictionary default");
                if (item is DxfXRecord record)
                    foreach (DxfTag tag in record.Data)
                        if (IsReference(tag)) handle((string)tag.Value, "XRECORD " + tag.Code);
                if (item is DxfOpaqueObject opaque)
                    foreach (DxfTag tag in opaque.Tags)
                        if (tag.ValueType == DxfTagValueType.Handle) handle((string)tag.Value, "opaque handle " + tag.Code);
                if (item is Section section) reference(section.GeometrySettings, "section settings");
                if (item is StoredTable table)
                    foreach (DxfObject target in table.References) reference(target, "ACAD_TABLE reference");
                if (item is MultiLeader leader)
                    foreach (MLeaderData data in leader.Data)
                        foreach (DxfObject target in data.References) reference(target, "MULTILEADER reference");
                if (item is Layout layout) reference(layout.PlotSettings?.ShadePlotObject, "layout shade plot 333");
            }
            foreach (HeaderVariable variable in this.Document.DrawingVariables.CustomValues())
            {
                DxfHandleKind kind = DxfGroupCode.GetHandleKind(variable.GroupCode);
                if (kind == DxfHandleKind.None || kind == DxfHandleKind.Arbitrary || variable.Name.Equals("$HANDSEED", StringComparison.OrdinalIgnoreCase)) continue;
                if (!(variable.Value is string value)) throw new InvalidOperationException("A custom header handle has an invalid value: " + variable.Name);
                if (handles.Contains(ErasureHandle(value))) throw ErasureReference(this.Document, "header " + variable.Name, value);
            }

            // No caller callbacks, graph discovery or validation occur after this point.
            // The internal registration removal event only releases APPID counts and subscriptions.
            foreach (DxfDictionaryEntry alias in aliases) parent.Remove(alias.Name);
            if (extension) root.Owner.ExtensionDictionary = null;
            foreach (DxfDatabaseObject item in tree)
            {
                this.Document.AddedObjects.Remove(item.Handle);
                this.objects.Remove(item.Handle);
                item.Database = null;
                item.IsErased = true;
            }
            root.Owner = null;
        }

        private List<DxfObject> ErasureCarriers()
        {
            var found = new HashSet<DxfObject>(ObjectIdentity);
            var result = new List<DxfObject>();
            Action<DxfObject> add = item => { if (item != null && found.Add(item)) result.Add(item); };
            foreach (DxfObject item in this.Document.AddedObjects.Values)
            {
                add(item);
                if (item is Insert insert) foreach (netDxf.Entities.Attribute attribute in insert.Attributes) add(attribute);
                if (item is Block block) add(block.End);
                if (item is Layout layout) add(layout.Viewport);
            }
            foreach (DxfDatabaseObject item in this.objects.Values) add(item);
            return result;
        }
        private static ulong ErasureHandle(string handle)
        {
            if (handle == null || handle.Length == 0 || handle.Length > 16 ||
                !ulong.TryParse(handle, NumberStyles.AllowHexSpecifier, CultureInfo.InvariantCulture, out ulong value))
                throw new InvalidOperationException("A retained handle cannot be safely inspected for erasure: " + handle);
            return value;
        }
        private static InvalidOperationException ErasureReference(DxfObject source, string field, string handle)
        { return new InvalidOperationException("Erasure would invalidate " + source.CodeName + " " + source.Handle + " " + field + " referencing " + handle + "."); }
    }
}

File: javascript/netDxf/Objects/DxfObjectDatabase.Erase.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import * as api from '../../index.js';
import { DatabaseRegistry } from '../../runtime/RegisteredDatabaseState.js';
import { OrdinalIgnoreCaseEquals } from '../../runtime/Collections.js';
import { ArgumentException, ArgumentNullException, InvalidOperationException, NotSupportedException } from '../../runtime/Errors.js';
const protectedSchemas = new Map([
  ['DxfStoredTableContent','Stored TABLECONTENT erasure requires its complete application schema.'],
  ['DxfStoredSunStudy','Stored SUNSTUDY erasure requires its complete application lifecycle.'],
  ['DxfStoredTableGeometry','Stored TABLEGEOMETRY erasure requires its complete application schema.'],
  ['DxfStoredCellStyleMap','Stored CELLSTYLEMAP erasure requires its complete application schema.'],
  ['DxfStoredField','Stored FIELD erasure requires its complete evaluator graph schema.'],
  ['DxfStoredDimAssoc','Stored DIMASSOC erasure requires the complete dimension association lifecycle.'],
]);
export function ErasureHandle(handle) {
  if (typeof handle !== 'string' || !/^[0-9a-f]{1,16}$/i.test(handle)) throw new InvalidOperationException('A retained handle cannot be safely inspected for erasure: ' + handle);
  return BigInt('0x' + handle);
}
export function ErasureReference(source, field, handle) { return new InvalidOperationException('Erasure would invalidate ' + source.CodeName + ' ' + source.Handle + ' ' + field + ' referencing ' + handle + '.'); }
export function InstallDatabaseErasure(Type) {
  Type.prototype.ErasureCarriers = function() {
    const found = new Set(), result = [], add = item => { if (item !== null && !found.has(item)) { found.add(item); result.push(item); } };
    for (const item of this.Document.AddedObjects.Values) {
      add(item);
      if (item instanceof api.Insert) for (const attribute of item.Attributes) add(attribute);
      if (item instanceof api.Block) add(item.End);
      if (item instanceof api.Layout) add(item.Viewport);
    }
    for (const item of DatabaseRegistry(this).Values) add(item);
    return result;
  };
  Type.prototype.EraseOwnedTree = function(root) { this.EraseOwnedTreeCore(root, null); };
  Type.prototype.EraseOwnedTreeCore = function(root, selectedManager) {
    if (root == null) throw new ArgumentNullException('root');
    if (root.IsErased) throw new InvalidOperationException('The object has already been erased.');
    if (root.Database !== this) throw new ArgumentException('The object must belong to this database.', 'root');
    this.CheckRegistered(root);
    if (root === this.Root) throw new InvalidOperationException('The named object dictionary root cannot be erased.');
    const carriers = this.ErasureCarriers(), children = new Map();
    for (const item of carriers) if (item.Owner !== null) { if (!children.has(item.Owner)) children.set(item.Owner, []); children.get(item.Owner).push(item); }
    const deleted = new Set(), tree = [], pending = [root];
    for (let at = 0; at < pending.length; at++) {
      const item = pending[at];
      if (deleted.has(item)) throw new InvalidOperationException('The erased ownership subtree contains a cycle.');
      deleted.add(item);
      if (!(item instanceof api.DxfDatabaseObject) || item === this.Root) throw new NotSupportedException('An ownership subtree containing a managed legacy object cannot be erased.');
      if (item.IsErased || item.Database !== this || !this.IsRegistered(item)) throw new InvalidOperationException('The erased ownership subtree has inconsistent registration.');
      for (const [name, message] of protectedSchemas) if (api[name] && item instanceof api[name]) throw new NotSupportedException(message);
      if (api.DxfStoredSectionManager && item instanceof api.DxfStoredSectionManager && item !== selectedManager) throw new NotSupportedException('Stored section-manager erasure requires the explicit manager lifecycle API.');
      if (item instanceof api.DxfOpaqueObject) throw new NotSupportedException('An opaque object requires its application schema before erasure: ' + item.CodeName);
      tree.push(item); if (children.has(item)) pending.push(...children.get(item));
    }
    if (root.Owner === null || !this.IsRegistered(root.Owner)) throw new InvalidOperationException('The erased root has no registered owner.');
    if (root.Owner === this.Document.Layers) throw new NotSupportedException('The LAYER table extension is managed by the layer-state collection.');
    const parent = root.Owner instanceof api.DxfDictionary ? root.Owner : null;
    const aliases = parent === null ? [] : Array.from(parent.Entries).filter(e => e.Target === root);
    if (parent === this.Root && aliases.some(e => Type.IsReservedName(e.Name))) throw new NotSupportedException('Managed legacy dictionary entries cannot be erased through this API.');
    const extension = root.Owner.ExtensionDictionary === root, sun = api.SunReferences.Get(root.Owner) === root;
    if (parent === null && !extension && !sun) throw new NotSupportedException('Erase an object through its owning dictionary or reciprocal extension attachment.');
    const handles = new Set();
    for (const item of tree) {
      const handle = ErasureHandle(item.Handle);
      if (handles.has(handle)) throw new InvalidOperationException('The erased ownership subtree contains duplicate numeric handles.');
      handles.add(handle);
      for (const data of item.XData.Values) if (!this.Document.ApplicationRegistries.References.ContainsKey(data.ApplicationRegistry.Name)) throw new InvalidOperationException('An erased object has inconsistent APPID reference bookkeeping: ' + data.ApplicationRegistry.Name);
    }
    for (const item of carriers) {
      if (deleted.has(item)) continue;
      const reference = (target, field) => { if (target !== null && deleted.has(target)) throw ErasureReference(item, field, target.Handle); };
      const handle = (target, field) => { if (handles.has(ErasureHandle(target))) throw ErasureReference(item, field, target); };
      reference(item.Owner, 'owner');
      if (!(sun && item === root.Owner)) reference(api.SunReferences.Get(item), 'SUN owner slot 361');
      if (!(extension && item === root.Owner && item.ExtensionDictionary === root)) reference(item.ExtensionDictionary, 'extension dictionary');
      for (const reactor of item.PersistentReactors) reference(reactor, 'persistent reactor');
      if (item instanceof api.EntityObject) for (const reactor of item.Reactors) reference(reactor, 'entity reactor');
      for (const data of item.XData.Values) for (const tag of data.XDataRecord) if (tag.Code === api.XDataCode.DatabaseHandle) handle(tag.Value, 'XData 1005');
      if (item instanceof api.DxfDatabaseObject) for (const target of item.DatabaseReferences) reference(target, 'typed reference');
      if (item instanceof api.DxfDictionary) for (const entry of item.Entries) if (!(item === parent && entry.Target === root)) reference(entry.Target, 'dictionary entry ' + entry.Name);
      if (item instanceof api.DxfDictionaryWithDefault) reference(item.Default, 'dictionary default');
      if (item instanceof api.DxfXRecord) for (const tag of item.Data) if (Type.IsReference(tag)) handle(tag.Value, 'XRECORD ' + tag.Code);
      if (item instanceof api.DxfOpaqueObject) for (const tag of item.Tags) if (tag.ValueType === api.DxfTagValueType.Handle) handle(tag.Value, 'opaque handle ' + tag.Code);
      if (item instanceof api.Section) reference(item.GeometrySettings, 'section settings');
      for (const [name, prefix] of [['Polyline3DRecord','polyline'], ['PolygonMeshRecord','polygon mesh'], ['PolyfaceMeshRecord','polyface'], ['Polyline2DRecord','legacy 2D']]) {
        if (item instanceof api[name]) { for (const target of item.References) reference(target, prefix + ' record reference'); for (const tag of item.OpaqueHandleTags) handle(tag.Value, prefix + ' record handle'); }
      }
      if (item instanceof api.PolyfaceMesh) for (const tag of item.StoredHeaderReferences) handle(tag.Value, 'polyface header handle');
      if (api.DxfOpaqueEntity && item instanceof api.DxfOpaqueEntity) for (const target of item.References) reference(target, 'unknown entity reference');
      if (item instanceof api.Polyline2D) for (const tag of item.StoredHeaderReferences) handle(tag.Value, 'legacy 2D header handle');
      if (api.StoredTable && item instanceof api.StoredTable) for (const target of item.References) reference(target, 'ACAD_TABLE reference');
      if (item instanceof api.MultiLeader) for (const data of item.Data) for (const target of data.References) reference(target, 'MULTILEADER reference');
      if (item instanceof api.Layout) reference(item.PlotSettings?.ShadePlotObject ?? null, 'layout shade plot 333');
    }
    for (const variable of this.Document.DrawingVariables.CustomValues()) {
      const kind = api.DxfGroupCode.GetHandleKind(variable.GroupCode);
      if (kind === api.DxfHandleKind.None || kind === api.DxfHandleKind.Arbitrary || OrdinalIgnoreCaseEquals(variable.Name, '$HANDSEED')) continue;
      const value = variable.Value instanceof api.BoxedString ? variable.Value.Value : variable.Value;
      if (typeof value !== 'string') throw new InvalidOperationException('A custom header handle has an invalid value: ' + variable.Name);
      if (handles.has(ErasureHandle(value))) throw ErasureReference(this.Document, 'header ' + variable.Name, value);
    }
    for (const alias of aliases) parent.Remove(alias.Name);
    if (extension) root.Owner.ExtensionDictionary = null;
    if (sun) api.SunReferences.Set(root.Owner, null, false);
    for (const item of tree) { this.Document.AddedObjects.Remove(item.Handle); DatabaseRegistry(this).Remove(item.Handle); item.Database = null; item.IsErased = true; }
    root.Owner = null;
  };
}
